site/src/examples/omezarr/omezarr-via-priority-cache.tsx

```diff
@@ -0,0 +1,291 @@
+import {
+  FancySharedCache,
+  logger,
+  type Resource,
+  type CachedTexture,
+  type WebResource,
+} from '@alleninstitute/vis-core';
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { pan, zoom } from '../common/camera';
+import REGL from 'regl';
+import { multithreadedDecoder } from '../common/loaders/ome-zarr/sliceWorkerPool';
+type DemoOption = { value: string; label: string; res: WebResource };
+
+class Tex implements Resource {
+  texture: CachedTexture;
+  constructor(tx: CachedTexture) {
+    this.texture = tx;
+  }
+  destroy() {
+    this.texture.texture.destroy();
+  }
+  sizeInBytes() {
+    return this.texture.bytes;
+  }
+}
+type Thing = {
+  tile: VoxelTile;
+  dataset: OmeZarrMetadata;
+  settings: RenderSettings;
+};
+function mapValues<T extends Record<string, V>, V, R>(obj: T, fn: (v: V) => R): { [k in keyof T]: R } {
+  return Object.keys(obj).reduce((acc, k) => {
+    return { ...acc, [k]: fn(obj[k]) };
+  }, {} as { [k in keyof T]: R });
+}
+
+function buildConnectedRenderer(regl: REGL.Regl, cache: FancySharedCache, decoder: Decoder, onData: () => void) {
+  //@ts-expect-error
+  const renderer = buildOmeZarrSliceRenderer(regl, decoder);
+  const client = cache.registerClient<Thing, Record<string, Tex>>({
+    cacheKeys: (item) => {
+      const channelKeys = Object.keys(item.settings.channels);
+      return channelKeys.reduce((chans, key) => {
+        return { ...chans, [key]: renderer.cacheKey(item.tile, key, item.dataset, item.settings) };
+      }, {});
+    },
+    fetch: (item) => {
+      // the Renderer<...> type obscures the fact that these are always cached textures... TODO  fix that?
+      // for now, this typecast is legit
+      const channels = renderer.fetchItemContent(item.tile, item.dataset, item.settings) as Record<
+        string,
+        (sig: AbortSignal) => Promise<CachedTexture>
+      >;
+      return mapValues(channels, (v: (sig: AbortSignal) => Promise<CachedTexture>) => {
+        return (sig: AbortSignal) => v(sig).then((tex) => new Tex(tex));
+      });
+    },
+    isValue: (v): v is Record<string, Tex> =>
+      renderer.isPrepared(
+        mapValues(v, (tx: Resource | undefined) => (tx && tx instanceof Tex ? tx.texture : undefined))
+      ),
+    onDataArrived: onData,
+  });
+  return {
+    render: (target: REGL.Framebuffer2D | null, dataset: OmeZarrMetadata, settings: RenderSettings) => {
+      const items = renderer.getVisibleItems(dataset, settings);
+      const baselayer = renderer.getVisibleItems(dataset, {
+        ...settings,
+        camera: { ...settings.camera, screenSize: [1, 1] },
+      });
+      regl.clear({ framebuffer: target, color: [0, 0, 0, 1], depth: 1 });
+      client.setPriorities(
+        new Set(items.map((tile) => ({ tile, dataset, settings }))),
+        new Set(baselayer.map((tile) => ({ tile, dataset, settings })))
+      );
+      for (const tile of [...baselayer, ...items]) {
```

Comment:
THIS is where it clicked for me.

With this sort of flexibility we should be able to get rid of the current major pixelation on zoom because we can always make sure to draw the existing zoomed out layer first while the others are loading, correct?

Reply:
yup - flexibility is exactly the goal - you're still stuck if you dont have what you want in the cache, but other than that you can do whatever you want - you could draw dots only if you have the pixels that go with them, you could draw some, clear the depth buffer, go get coffee, then come back and render the dots - whatever you can think of!

Reply:
I think my current plan (for that pixelation issue) would be to keep a list of tiles drawn recently - and use them to fill in holes while we wait for new stuff to load. In addition, this makes it easy to pre-fetch things - we could pre-fetch slices N-1 and N+1 simply by listing them as low-priority (and then simply not drawing them!)